src/Billing/AutoBilling.php

```diff
@@ -0,0 +1,278 @@
+<?php
+
+/*
+ *  package   OpenEMR
+ *  link      http://www.open-emr.org
+ *  author    Sherwin Gaddis <sherwingaddis@gmail.com>
+ *  copyright Copyright (c )2020.. Sherwin Gaddis <sherwingaddis@gmail.com>
+ *  license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+ */
+
+namespace OpenEMR\Billing;
+
+use Exception;
+
+class AutoBilling
+{
+    public function __construct()
+    {
+        //do epic stuff...
+    }
+
+    public function generateBilling($event_date, $bInfo, $pid, $userid)
+    {
+        /**
+         *  Aggregate information to fill billing table
+         */
+        $event_date = $event_date . " " . date("H:i:s"); // append current time to event date
+        $billingData = explode("-", $bInfo);  //separate up billing information into code and text
+        if (!empty($billingData[0])) {
+            $code = trim($billingData[0]);  // code that is in the calendar description
+            $text = trim($billingData[1]);  // text that is after the code in the calendar description
+        } else {
+            echo xlt("Calendar entry not properly formated");
+            die;
+        }
+        //see if there is a diagnosis in the patient chart
+        //returns an array
+        $icd10 = $this->getDiagnosis($pid);
```

Comment:
next step is to make these variable part of the class. For example, will have:
```
$this->pid = $pid;
```
then when call the function do:
```
$icd10 = $this->getDiagnosis();
```
then inside the getDiagnosis() function, use the `$this->pid` rather than $pid.

Reply:
and then at top of class will have for this variable:
```
private $pid;
```

Reply:
the reason the $pid is sent in rather than referenced by the session is that the user will not be in the patient chart. The user will be in the calendar only closing the appointment.  So, the user is not forced to open the chart to set the PID variable in the session. If the appointment is being closed from the calendar the PID is a post variable. The only thing I can think of is that @bradymiller is suggesting that the variable be set this way.

Inside the add edit event do this:

            $doBilling->pid = $_POST['pid']
         $doBilling->bInfo = $_POST['form_title']
        $doBilling->userid = $_SESSION['authUserID']

Inside AutoBilling class do this

      class AutoBilling
      {
          private $pid;
          private $bInfo;
          private $userid;
          private $event_date;

The reason that the credentials are in the globals is that I was thinking to make it generic enough that any clearinghouse providing the SFTP information to their client. The client could plug in the information. It is not meant to be specific to a clearinghouse. Also, most practice use only one clearinghouse in my experience. But if there is a practice that uses more than one clearinghouse at a time. It should be moved to practice settings.

The upload should be async. If it is not set that way. I did not find any setting here http://phpseclib.sourceforge.net/sftp/2.0/examples.html#put

Reply:
Nothing wrong with SFTP as a protocol. I believe that logically and functionally, the credentials belong with the x-12 partner. Dan already has code for that. 

To make the transfer asynchronous, it would need to be a background task. I would propose putting references to files in a table, as a queue, and processing transfers from the queue rather than doing them synchronously based on a user action. That way new files, as well as files that haven’t been sent successfully can be tracked and processed. 

Dan and I are wrapping up very similar functionality that we were going to commit this week, so this stuff is on the top of my brain. @growlingflea

Reply:
Maybe we can chat this week on how we can coordinate our effort since it sound like we’re working on similar things?

Reply:
@juggernautsei ,
I was just pointing out that might as well make the variables declared within your class since you are calling private functions (so they can just use variables declared within the class rather than passing as parameters). Not suggesting use of globals or sessions.

Reply:
@juggernautsei , also to clarify, this is a minor issue and no need to pursue this especially since looks like more vital stuff to focus on per @kchapple .

Reply:
@kchapple we can talk this week. I have a 10 AM meeting on Tuesday. I am free all day Monday (tomorrow). I am free all day Wednesday. Let me know what time fits into your schedule.

Reply:
hi guys, could I listen in as well please? we could use the [openemr zoom setup](https://www.open-emr.org/wiki/index.php/OEMR_Monthly_Conference_Call_Details)

thank you

Reply:
Not sure if current code still has it but 'Electronic Report' checked all configured SFTP laboratory connections every time that screen was called. This would cause delays even if the user was just looking for history. SFTP connections need a decent timeout so any down hosts would cause major wait.

Ended up extracting fetch and update code in a separate script for use by cron AND added a button to run the same when the user was ok with expected short delay as results needed to be fetched stat.

Reply:
@mdsupport in this workflow. The SFTP is only activated when there is a file to send. If no file exists, then the SFTP call is not made to put the file on the remote server. Also, the users' action has to be purposeful to transmit a claim. Without the user selecting the transmit of the claim the sFTP is not fired. 

@mdsupport, I looked it up and to add a time out to the sFTP. All I have to add is 

     $ssh->setTimeout(20);

In your experience, is 20 seconds a good time out for the login to fail or go through? I have pushed the commit for that update.

Reply:
@juggernautsei 20s for clearinghouse should be more than adequate. 

I was just narrating our user experience years ago when labs interface was activated. Without knowing the billing workflow my guess was clearinghouse was not that different from labs - get inbound and send outbound. Obviously that does not seem to be the case for billing.

Best.

Reply:
@kchapple How is your project coming along? I paused work on this because you stated that you have a similar project that you are working on. Have you posted your project?